_posts/2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,253 @@
+Dafny already migrates those icons according to string edits.
+I also found the lightning metaphor to be adequate to provide the sense of "verification" on the side of the vision field without hiding the previous verification status.
+However, I observed that if we just took the rectangle for error and put a lightning on it (yellow or transparent), it would just not be visible to the side of the vision field.
+This is especially more important given that I modified Dafny to give per-method feedback.
+This is why the red rectangle is split in screenshots E and F of [Figure 2](#fig_dynamic_gutter_icons), as if the error was about to "go away".
+
+Among all the dynamic icons, note that the topmost part of [Figure 2](#fig_dynamic_gutter_icons) is less distinct than for code in error context. Without more information about the new code, I preferred the dynamic icons to reflect a kind of default hypothesis that previously verified code will probably also be verified.
+I wanted Dafny to feel **more helpful than inquisitive**, i.e. **optimistic about the ability for the user to write correct code**.
+
+I found it was appropriate to refresh these icons every time a declaration changes its verification status, in real-time. Moreover, in the current absence of a reliable caching mechanism, I added an edit detection mechanism to ensure that the last methods being edited are verified first.
+In [Figure 2](#fig_dynamic_gutter_icons), verification being done on more than 2 cores, it was possible for the verification of the lines 1-7 to finish before the others.
+
+
+#### Implementation details and cosmetics
+
+<!--Things get deep now. Feel free to take a break if you really want to read this, or [skip to the next section](#enhancing-positive-and-negative-feedback) to learn more about things other than gutter icons.-->
+
+To render static and dynamic icons, I used a z-buffer mechanism with clever codes for each icons.
+I also have had special icons designed for the start and end of declarations with errors so that if flows nicely with verified code (for example <img src="/blog/assets/images/verification-compelling/example1.png" alt="Dafny gutter icons on dark theme" width="45%" style="width:25px;height:24px;object-fit:none;object-position:-7px -307px;margin:3px"/>)
+In real-time feedback, I also delayed the display of gutter icons by 2 seconds in the presence of parse or resolution errors, to avoid icons blinking when typing fast.
+Finally, I ensured that the scroll bar itself reflects gutter icons, so that it's possible to see at a glance if the entire file is verified, and if not, where to scroll.
+
+<div id="enhancing-positive-and-negative-feedback"></div>
+
+## Enhanced positive and negative hover feedback
+
+Congratulations for reading until here! So far, you already have a sense on how the interface of Dafny is almost gamified with its on heads-up display or HUD, which contributes to making it delightful.
```

Comment:
nit: I'm not sure I would think of this interface as a HUD, but it's certainly an immersive UI experience.